lib/libuzfs/libuzfs.c

```diff
@@ -1932,6 +1943,75 @@ libuzfs_object_write(libuzfs_inode_handle_t *ihp, uint64_t offset,
 	    iovs, iov_cnt, sync, 0));
 }
 
+int libuzfs_object_read_zero_copy(libuzfs_inode_handle_t *ihp,
+    uint64_t offset, uint64_t size, libuzfs_read_buf_t *buf)
+{
+	ASSERT3U(size, <=, DMU_MAX_ACCESS / 2);
+
+	zfs_locked_range_t *lr = zfs_rangelock_enter(&ihp->rl,
+	    offset, size, RL_READER);
+	ASSERT(lr != NULL);
+	bzero(buf, sizeof (*buf));
+	if (offset >= ihp->u_size) {
+		zfs_rangelock_exit(lr);
+		return (0);
+	}
+
+	uint64_t nread = MIN(ihp->u_size - offset, size);
+
+	dmu_buf_impl_t *db = (dmu_buf_impl_t *)sa_get_db(ihp->sa_hdl);
+	DB_DNODE_ENTER(db);
+	dnode_t *dn = DB_DNODE(db);
+	int err = dmu_buf_hold_array_by_dnode(dn, offset, nread,
+	    TRUE, NULL, &buf->num_bufs, &buf->dbpp, 0);
+	DB_DNODE_EXIT(db);
+	if (unlikely(err != 0)) {
+		zfs_rangelock_exit(lr);
+	} else {
+		libuzfs_inode_handle_inc_ref(ihp);
+		buf->ihp = ihp;
+		buf->lr = lr;
+		buf->offset = offset;
+		buf->nread = nread;
+	}
+
+	return (err);
+}
+
+void
+libuzfs_read_buf_to_slices(const libuzfs_read_buf_t *read_buf,
+    libuzfs_slices_t *slices)
+{
+	uint64_t offset = read_buf->offset;
```

Comment:
how to handle hole in file?

Reply:
if a dbuf is a hole,  dmu_buf_hold_array_by_dnode will fill it with zero,